Database backup contrib to schedule database backups into /server/backups

aMiss-aWry opened the pull request:
#### Brief overview of PR changes/additions
New `backup` command, which starts/manages a `DatabaseBackupScript`, which runs daily by default.

The script saves a copy of the database to the game `server/backups` folder. It uses `shutil.copy` for sqlite3 and `subprocess.run` `pg_dump` for postgresql. This PR only backs the db up to the same filesystem; users should decide what they want to do next (upload to the cloud, download to a hard drive etc).

#### Motivation for adding to Evennia
I needed to write it for myself and could imagine others finding it useful.

#### Other info (issues closed, discussion etc)
Only supports sqlite3 and postgresql at the moment.

#### Screengrabs

Sqlite3:
![image](https://github.com/user-attachments/assets/13e40b57-3917-4ba2-a4bf-0f4a535a7d10)

In folder:
```
(evenv)  (master *) backups > pwd
[..]/desktop/muddev/mygame/server/backups

(evenv)  (master *) backups > ls
2024-10-15T21:05:06.db3	2024-10-15T21:05:10.db3
```

PostgreSQL:
![image](https://github.com/user-attachments/assets/66f49964-de4d-49ad-9329-c34b2c6f4e15)

In folder:
```
(evenv)  (master *) backups > pwd
[...]/desktop/muddev/gardenmud/server/backups
(evenv)  (master *) backups > ls
2024-10-15T21:06:55.sql	2024-10-15T21:07:02.sql
```

aMiss-aWry committed:
Database backup contrib with CmdBackup

aMiss-aWry committed:
Update docs

aMiss-aWry committed:
Added tests, cleaned up messaging

aMiss-aWry committed:
Fix some docstrings

Griatch commented on evennia/contrib/utils/database_backup/README.md:
```diff
@@ -0,0 +1,69 @@
+
+### Restoring sqlite3 (.db3)
+
+* Copy the database backup you want to restore from back into the `server/` directory
+* Rename the database backup to `evennia.db3` if you have not modified `settings.DATABASES`, otherwise whatever name is in your `settings.DATABASES` dictionary.
```
If you don't know that the sqlite3 file is named `evennia.db3`, this instruction becomes a little confusing. I think it would be clearer if this was phrased something like
```
By default (unless you changed the name of the file in `settings DATABASES`), the sqlite3 file `mygame/server/evennia.db3`. Copy your backup file over this file to recover your backup.
```

aMiss-aWry replied:
Fixed wording according to suggestion

Griatch commented on evennia/contrib/utils/database_backup/README.md:
````diff
@@ -0,0 +1,69 @@
+
+* Prepare the following variables
+```
+export DB_USER=db_user from your settings.DATABASES
+export DB_NAME=db_name from your settings.DATABASES
+export BACKUP_FILE=the path to the backup file you are restoring from
+export PGPASSWORD=your db password
````
This is hard to understand since it's all in \`\`\`. Is `export DB_USER=db_user from your settings.DATABASES` one command? Is `export DB_USER=db_user` the command you should run? Does 'export' here only mean you should find it and remember it? If you use \`\`\`, place the actual command alone (with comments between each block), or use clear `#`-markers to show what is a comment.

Griatch commented on evennia/contrib/utils/database_backup/README.md:
````diff
@@ -0,0 +1,69 @@
+
+Currently, the sqlite3 (the evennia default) and postgresql databases are supported. Others are welcome to add more.
+
+## Installation
+
+This utility adds the `backup` command. Import the module into your commands and add it to your command set to make it available.
+
+In `mygame/commands/default_cmdsets.py`:
+
+```python
+...
+from evennia.contrib.utils.database_backup import DbCmdSet   # <---
+
+class CharacterCmdset(default_cmds.Character_CmdSet):
+    ...
+    def at_cmdset_creation(self):
+        ...
+        self.add(DbCmdSet)  # <---
+
+```
+
+Then `reload` to make the `backup` command available.
````
It installs the backup command, but talks about a script. It doesn't actually describe how this is to be used - do you use the backup command to do the backup? Does it start a script? What? You shouldn't have to go read the code to find out.

Griatch commented on evennia/contrib/utils/database_backup/README.md:
```diff
@@ -0,0 +1,69 @@
+
+## Installation
+
+This utility adds the `backup` command. Import the module into your commands and add it to your command set to make it available.
```
Since this is based on a global Script, why not also describe how to just add the script as a global script in settings; that way it could be set up without having to connect to the game and run a command ...

Griatch commented on evennia/contrib/utils/database_backup/database_backup.py:
```diff
@@ -0,0 +1,184 @@
+            _MUDINFO_CHANNEL.msg(message)
+        logger.log_sec(message)
+
+    def backup_postgres(self, db_name, db_user, output_file_path):
+        """
+        Run `pg_dump` on the postgreSQL database and save the output.
+        """
+
+        output_file_path += ".sql"
+        subprocess.run(
+            ["pg_dump", "-U", db_user, "-F", "p", db_name],
+            stdout=open(output_file_path, "w"),
+            check=True,
+        )
+        self.log(f"|wpostgresql db backed up in: {BACKUP_FOLDER}|n")
+
+    def backup_sqlite3(self, db_name, output_file_path):
+        """
+        Copy the sqlite3 db.
+        """
+
+        output_file_path += ".db3"
+        os.makedirs(os.path.dirname(output_file_path), exist_ok=True)
+        shutil.copy(db_name, output_file_path)
+        self.log(f"|wsqlite3 db backed up in: {BACKUP_FOLDER}|n")
```
How safe is this operation? The sqlite3 file is just a binary and `shutil.copy` doesn't know anything about what a database is. What happens if this is being written to (by an active game) while this copy operation is performed?

aMiss-aWry replied:
I think probably the easiest thing to do here is add a `PRAGMA integrity_check` post-op and log a failure if it doesn't work, indicating the backup was corrupted and needs to be retaken. Thoughts?

aMiss-aWry replied:
Did the PRAGMA integrity_check thing.

Griatch reviewed (changes requested):
This is a cool contrib I think people will find useful! Some comments: 

- I made some clarifications suggestions for the README
- It's nice to have a command, but I imagine a lot of people would be happy to just add the global script to settings with some sane defaults without needing to run a command; seems like a good thing to add to readme and to consider as a use case! 
- The contrib's unit tests all fail. Most likely you tested this in isolation, but is not cleaning up the twisted deferreds of the script (I would suggest mocking instead of actually running a timed operation as part of a unit test).

Griatch commented:
Ping.

Griatch commented:
Ping :)

aMiss-aWry committed:
Merge branch 'evennia:main' into database_backup_contrib

aMiss-aWry committed:
Updated README per comments

aMiss-aWry committed:
Add PRAGMA integrity_check post-copying sqlite3 db

Griatch commented:
@aMiss-aWry The unit tests are failing across the board here.

Griatch commented on evennia/contrib/utils/database_backup/__init__.py:
```diff
@@ -0,0 +1,5 @@
+Database backups - helpme 2024
+"""
+
+from .database_backup import DbCmdSet  # noqa
```
You are supposed to use `DatabaseBackupScript` directly from this module I think. So it should probably be added here too.

aMiss-aWry replied:
Done

Griatch commented on evennia/contrib/utils/database_backup/tests.py:
```diff
@@ -0,0 +1,88 @@
+    @patch("evennia.utils.logger.log_sec")
+    def test_sqlite_success(self, mock_logger, mock_copy):
+        mock_copy.return_value.returncode = 0
+        self.call(
```
This is actually creating a script in the database afaik, and failures we see would be due to leaving Twisted deferred's dangling outside of the test scope. Either mock `create_script` or be careful to clean up in `tearDown`.

aMiss-aWry replied:
Added to teardown.

Griatch commented on evennia/contrib/utils/database_backup/database_backup.py:
```diff
@@ -0,0 +1,195 @@
+import sqlite3
+
+_MUDINFO_CHANNEL = None
+BACKUP_FOLDER = "server/backups"
```
This works, but it's more os-agnostic to use `os.path.join(settings.GAME_DIR, "server", "backups")` here.

aMiss-aWry replied:
Done

Griatch commented on evennia/contrib/utils/database_backup/database_backup.py:
```diff
@@ -0,0 +1,195 @@
+import os
+import shutil
+import subprocess
+import sqlite3
```
Is this actually used anywhere?

aMiss-aWry replied:
It looks like you commented this on the import for os, shutil, subprocess, sqlite3 - yes they're all used in the script. If this was meant to be on the `from evennia.comms.models import ChannelDB` - yes that was unused, removed.

Griatch reviewed (changes requested):
@aMiss-aWry 
Looking back at this one (doing a lot of PR reviewing today), there are a few things that need to happen for this PR to merge: 

- See inline comments 
- The `postgresql_psycopg2` is `postgresql` these days.

aMiss-aWry committed:
Merge branch 'evennia:main' into database_backup_contrib

aMiss-aWry committed:
Fixing tests - mocking sqlite3 setting, tearing down script

Using DATETIME_FORMAT_STR instead of isoformat(), Windows dislikes
colons in filenames.
Removing unused import.
Adding script to __init__.py

aMiss-aWry committed:
Fixing tests - mocking sqlite3 setting, tearing down script

Using DATETIME_FORMAT_STR instead of isoformat(), Windows dislikes
colons in filenames.
Removing unused import.
Adding script to __init__.py
Added docstring
Added database check in at_pre_cmd()
Remove corrupt backups if found
Use sqlite3 .backup instead of shutil copy
Filter on extension to find backup files

Griatch commented on evennia/contrib/utils/database_backup/database_backup.py:
```diff
@@ -0,0 +1,202 @@
+from django.conf import settings
```
As per CODESTYLE, the module needs a docstring describing what it's doing.

Griatch commented on evennia/contrib/utils/database_backup/README.md:
```diff
@@ -0,0 +1,84 @@
+Contribution by helpme (2024)
+
+This module schedules backups in-game, which saves a copy of your database to your game's `server/backups` folder. Database backups are *not* automatically uploaded to any cloud service, it is left to you to decide what to do with them (i.e. pushed to git, uploaded to the ether, downloaded to a hard drive).
+
```
It should be mentioned that MySQL/MariaSQL databases are not supported here.

aMiss-aWry replied:
The supported dbs are mentioned in the next couple lines

Griatch commented on evennia/contrib/utils/database_backup/database_backup.py:
```diff
@@ -0,0 +1,202 @@
+        args = self.args.strip()
+        interval = int(args) if args.isnumeric() else DEFAULT_INTERVAL
+
+        script = self.get_script()
```
I think there needs to be a check for which database is configured, and the command aborting if finding a non-supported one (MySQL/MariaDB). Otherwise a user may run this, get a success message and then `at_repeat` will silently skip doing the backup every repeat, which is not great ...

aMiss-aWry replied:
Added check in at_pre_cmd()

Griatch commented on evennia/contrib/utils/database_backup/database_backup.py:
```diff
@@ -0,0 +1,202 @@
+        con = sqlite3.connect(output_file_path)
+        cur = con.cursor()
+        try:
+            cur.execute("PRAGMA integrity_check")
```
If this fails, the corrupt db3 file is left on disk - and will be counted as "latest backup". This gives a false sense of security. It seems better to delete the corrupted or move it to a `*_corrupted` suffix so the backup is recreated later.

aMiss-aWry replied:
Fixed - deleting corrupt backup

Griatch commented on evennia/contrib/utils/database_backup/database_backup.py:
```diff
@@ -0,0 +1,202 @@
+
+        output_file_path += ".db3"
+        os.makedirs(os.path.dirname(output_file_path), exist_ok=True)
+        shutil.copy(db_name, output_file_path)
```
It may be worth pointing out the risks of this in the README. The Shutil solution is fast but could clash with writes (maybe sqlite library's inbuilt backup command may be worth exploring?)

aMiss-aWry replied:
Removed shutil - using sqlite's backup

Griatch commented on evennia/contrib/utils/database_backup/database_backup.py:
```diff
@@ -0,0 +1,202 @@
+        try:
+            files = os.listdir(BACKUP_FOLDER)
+            paths = [os.path.join(BACKUP_FOLDER, basename) for basename in files]
+            last_backup = max(paths, key=os.path.getctime)
```
Any file (like a README, tar file etc) becomes a 'backup' to this. While maybe unlikely, it's unexpected behavior, and pretty fragile. It would make sense to filter on `.db3` or `.sql` or whatever is needed.

aMiss-aWry replied:
Fixed - filtering on `.db3`/`.sql`

Griatch commented on evennia/contrib/utils/database_backup/database_backup.py:
```diff
@@ -0,0 +1,202 @@
+            self.create_script(interval)
+            return
+
+        if "force" in self.switches:
```
If the script didn't previously exist, the func() will have exited on line 172 before this is evaluated. That feels unexpected to the user.

aMiss-aWry replied:
Reordered

Griatch commented on evennia/contrib/utils/database_backup/database_backup.py:
```diff
@@ -0,0 +1,202 @@
+        """
+
+        output_file_path += ".sql"
+        with open(output_file_path, "w") as output_file:
```
This already creates the sql file on disk. What happens if pg_dump fails (bad creds, not installed etc)? You'll end up with an empty/corrupted backup file  (false sense of security).

aMiss-aWry replied:
Added cleanup in catch

Griatch reviewed (changes requested):
Thanks for the fixes. Looked things over on the functional side / played around, and found a few more things that is behaving strangely, or is not as safe as one would expect a backup to be.

Griatch reviewed (approved):
Thanks for addressing the feedback! 
The only caveat I can think of is that setting this up via `global_scripts` (as suggested in README) will bypass the db-check in `at_pre_cmd` (since no cmd is called), and thus unsupported engines will not be detected. But if you set things up like that it's probably a minor concern, so approving.
